File: src/main/java/me/zodiakk/spigotjs/engine/object/Block.java
package me.zodiakk.spigotjs.engine.object;

public interface Block extends ApiElement {
    public boolean breakNaturally();
    public boolean breakNaturally(ItemStack tool);
    public String getBiome();
    public Number getPower();
    public Number getPower(String face);
    public ItemStack[] getDrops();
    public ItemStack[] getDrops(ItemStack tool);
    public String getFace(Block block);
    public Number getHumidity();
    public Number getBlocksLightLevel();
    public Number getSkyLightLevel();
    public Number getLightLevel();
    public Location getLocation();
    public String getPistonReaction();
    public Block getRelative(String face);
    public Block getRelative(String face, Number distance);
    public Block getRelative(Number x, Number y, Number z);
    public Number getTemperature();
    public String getType();
    public World getWorld();
    public Number getX();
    public Number getY();
    public Number getZ();
    public boolean isBlockFaceIndirectlyPowered(String face);
    public boolean isBlockFacePowered(String face);
    public boolean isBlockIndirectlyPowered();
    public boolean isBlockPowered();
    public boolean isEmpty();
    public boolean isLiquid();
    public void setBiome(String biome);
    public void setType(String material);
    public void setType(String material, boolean applyPhysics);
}

File: src/doc/js/types/Block.ts
import { ItemStack, Location, World } from 'types';
import 'typedefs';

/**
 * Represents a single Block in the world.
 */
export abstract class Block {
    static noop: Block;

    breakNaturally(tool?: ItemStack): boolean { return false; }
    getBiome(): Biome { return 'plains'; }
    getPower(face?: BlockFace): number { return 0; }
    getDrops(item?: ItemStack): ItemStack[] { return []; }
    getFace(block: Block): BlockFace | null { return null; }
    getHumidity(): number { return 0; }
    getBlocksLightLevel(): number { return 0; }
    getSkyLightLevel(): number { return 0; }
    getLightLevel(): number { return 0; }
    getLocation(): Location { return Location.noop; }
    getPistonReaction(): PistonReaction { return 'ignore'; }
    getRelative(face: string, distance?: number): Block;
    getRelative(x: number, y: number, z: number): Block;
    getRelative(a0: string | number, a1: number | void, a2: number | void): Block { return Block.noop; }
    getTemperature(): number { return 0; }
    getType(): string { return ''; }
    getWorld(): World { return World.noop; }
    getX(): number { return 0; }
    getY(): number { return 0; }
    getZ(): number { return 0; }
    isBlockFaceIndirectlyPowered(face: BlockFace): boolean { return false; }
    isBlockFacePowered(face: BlockFace): boolean { return false; }
    isBlockIndirectlyPowered(): boolean { return false; }
    isBlockPowered(): boolean { return false; }
    isEmpty(): boolean { return false; }
    isLiquid(): boolean { return false; }
    setBiome(biome: Biome): void {}
    setType(type: string, applyPhysics?: boolean): void {}
}
